efi/preinstall/check_pcr7.go

```diff
@@ -0,0 +1,529 @@
+		{Name: "PK", GUID: efi.GlobalVariable},
+		{Name: "KEK", GUID: efi.GlobalVariable},
+		{Name: "db", GUID: efi.ImageSecurityDatabaseGuid},
+		{Name: "dbx", GUID: efi.ImageSecurityDatabaseGuid},
+		// TODO: Add optional dbt / SPDM in the future.
+	}
+
+	result = new(secureBootPolicyResult)
+	var (
+		db                        efi.SignatureDatabase // The authorized signature database from the TCG log.
+		measuredSignatures        tpm2.DigestList       // The verification event digests measured by the firmware
+		seenOSPresentVerification bool                  // Whether we've seen a verification event in the OS-present phase
+		seenIBLLoadEvent          bool                  // Whether we've seen the launch event for the OS initial boot loader
+	)
+
+	phaseTracker := newTcgLogPhaseTracker()
+NextEvent:
+	for _, ev := range log.Events {
+		phase, err := phaseTracker.processEvent(ev)
+		if err != nil {
+			return nil, err
+		}
+
+		switch phase {
+		case tcglogPhasePreOSMeasuringSecureBootConfig:
+			if ev.PCRIndex != internal_efi.SecureBootPolicyPCR {
+				// Not PCR7
+				continue NextEvent
+			}
+
+			switch ev.EventType {
+			case tcglog.EventTypeEFIVariableDriverConfig:
+				if len(configs) == 0 {
+					// Unexpected config event - we're not expecting another secure boot variable
+					// to measure. We should have exitted the loop by now.
+					return nil, errors.New("unexpected EV_EFI_VARIABLE_DRIVER_CONFIG event: all expected secure boot variable have been measured")
+				}
+
+				// Pop the next secure boot config name
+				config := configs[0]
+				configs = configs[1:]
+
+				data, ok := ev.Data.(*tcglog.EFIVariableData)
+				if !ok {
+					// The data resulting from decode errors are guaranteed to implement the error interface
+					return nil, fmt.Errorf("invalid event data for EV_EFI_VARIABLE_DRIVER_CONFIG event: %w", ev.Data.(error))
+				}
+				// Make sure this is the event we're expecting to be measured. If they're
+				// measured in an unexpected order, then WithSecureBootPolicyProfile() will
+				// generate an invalid policy.
+				if data.VariableName != config.GUID || data.UnicodeName != config.Name {
+					return nil, fmt.Errorf("unexpected EV_EFI_VARIABLE_DRIVER_CONFIG event ordering (expected %s-%v, got %s-%v)",
+						config.Name, config.GUID, data.UnicodeName, data.VariableName)
+				}
+
+				// Compute the expected digest from the event data in the log and make
+				// sure it's consistent with the measured digest.
+				expectedDigest := tcglog.ComputeEFIVariableDataDigest(pcrAlg.GetHash(), data.UnicodeName, data.VariableName, data.VariableData)
+				if !bytes.Equal(ev.Digests[pcrAlg], expectedDigest) {
+					return nil, fmt.Errorf("event data inconsistent with measured digest for EV_EFI_VARIABLE_DRIVER_CONFIG event (name:%q, GUID:%v, expected digest:%#x, measured digest:%#x)",
+						data.UnicodeName, data.VariableName, expectedDigest, ev.Digests[pcrAlg])
+				}
+
+				switch data.UnicodeName {
```

Comment:
this third level of switch is a bit much, I wonder if we can make an helper of this level of checks